File: chapter02/sanghyo/MovieReservationProgram/MovieReservationProgram/Model/Money.swift
//
//  Money.swift
//  MovieReservationProgram
//
//  Created by sanghyo on 11/29/23.
//

import Foundation

struct Money {
    static let ZERO: Money = Money.wons(0)
    
    private let amount: Float
    
    static func wons(_ amount: Int) -> Money {
        return Money(Float(amount))
    }
    
    static func wons(_ amount: Float) -> Money {
        return Money(amount)
    }
    
    init(_ amount: Float) {
        self.amount = amount
    }
    
    func plus(money: Money) -> Money {
        return Money(self.amount + money.amount)
    }
    
    func minus(money: Money) -> Money {
        return Money(self.amount - money.amount)
    }
    
    func times(percent: Float) -> Money {
        return Money(self.amount*percent)
    }
    
    func times(cnt: Int) -> Money {
        return Money(self.amount*Float(cnt))
    }
    
    func isLessThan(other: Money) -> Bool {
        return self.amount < other.amount
    }
    
    func isGreaterThanOrEqual(other: Money) -> Bool {
        return self.amount >= other.amount
    }
}

File: chapter02/hm4725/src/Money.cc
#include "Money.hh"
#include <memory>

/* public */
const std::shared_ptr<Money> Money::ZERO = Money::wons(0);

/* protected */
Money::Money(long amount) : amount(amount) {}

/* public */
std::shared_ptr<Money> Money::wons(long amount) {
  struct MakeSharedEnabler : public Money {
    MakeSharedEnabler(long amount) : Money(amount) {}
  };
  return std::make_shared<MakeSharedEnabler>(amount);
}

std::shared_ptr<Money> Money::getptr() { return shared_from_this(); }

std::shared_ptr<Money> Money::plus(std::shared_ptr<Money> money) {
  return wons(amount + money->amount);
}

std::shared_ptr<Money> Money::minus(std::shared_ptr<Money> money) {
  return wons(this->amount - money->amount);
}

std::shared_ptr<Money> Money::times(double percent) {
  return wons(this->amount * percent);
}

bool Money::isLessThan(std::shared_ptr<Money> money) {
  return this->amount < money->amount;
}

bool Money::isGreaterThanOrEqual(std::shared_ptr<Money> money) {
  return this->amount >= money->amount;
}

long Money::getAmount() {
  return amount;
}